### PR TITLE
Add R.thunk

### PR DESCRIPTION
**Pull request by raine**
Discussed in #1366

What is the correct signature? The current one is same as `R.partial`'s.

### CODE REVIEW

File: src/thunk.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * @return {Function} A new function wrapping `fn`. When invoked, it will call `fn`
+ *         with `args`, while disregarding any new arguments.
+ * @see R.partial
+ * @example
+ *
+ *      var add = (a, b) => a + b;
+ *      var sum = (...args) => args.reduce(add, 0);
+ *
+ *      var sumThunk = R.thunk(sum, 1, 2)
+ *      sumThunk(2); //=> 3 // 2 is disregarded
```
Comment (CrossEye) [inline src/thunk.js]: sumThunk(5); //=> 3 // 5 is disregarded

so as not to confuse the two 2's.

File: test/thunk.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+var eq = require('./shared/eq');
+
+describe.only('thunk', function() {
+  var add = function(a, b) { return a + b; };
+  var sum = function() {
+    return [].reduce.call(arguments, add, 0);
+  };
+
+  it('applies function only to arguments given initially', function() {
+    eq(R.thunk(sum, 1, 1)(2), 2);
```
Comment (CrossEye) [inline test/thunk.js]: ditto

**Comment by CrossEye** [comment]
:+1:

File: src/thunk.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+var partial = require('./partial');
+
+/**
+ * Accepts as its arguments a function and any number of values and returns a function that,
+ * when invoked, calls the original function with only the values originally provided.
+ *
+ * @func
+ * @memberOf R
+ * @category Function
+ * @sig (a -> b -> ... -> i -> j -> ... -> m -> n) -> a -> b -> ... -> i -> (j -> ... -> m -> n)
```
Comment (scott-christopher) [inline src/thunk.js]: I think the signature should show that this returns a nullary function.
e.g.
```
(a -> b -> ... -> i -> j) -> a -> b -> ... -> i -> (() -> j)
```

File: src/thunk.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * @return {Function} A new function wrapping `fn`. When invoked, it will call `fn`
+ *         with `args`, while disregarding any new arguments.
+ * @see R.partial
+ * @example
+ *
+ *      var add = (a, b) => a + b;
+ *      var sum = (...args) => args.reduce(add, 0);
+ *
+ *      var sumThunk = R.thunk(sum, 1, 2)
+ *      sumThunk(5); //=> 3 // 5 is disregarded
```
Comment (scott-christopher) [inline src/thunk.js]: While it is good to demonstrate extra arguments are ignored, I think this example should also include the more common use case of evaluating the thunk with no args.

Perhaps including two examples of the same thunk being called could be more informative.
```js
var sumThunk = R.thunk(sum, 1, 2)
sumThunk();  //=> 3
sumThunk(5); //=> 3 // 5 is disregarded
```

**Comment by raine** [comment]
Good points, @scott-christopher and @CrossEye

**Comment by raine** [comment]
The `compose(nAry(0), partial);` implementation wouldn't satisfy the invariants test so I changed the implementation.

```
invariants
  1) -- applying function f with length n (where n > 0) to no arguments gives function with length n
  2) -- applying function f with length n (where n > 0) to R.__ gives function with length n
  ✓ -- applying function f with length n (where n > 1) to any value other than R.__ gives function with length n - 1
```

Before:

```
> var thunk = require('./src/thunk')
> thunk.length
1
> thunk().length
0
```

After:

```
> var thunk = require('./src/thunk')
> thunk.length
1
> thunk().length
1
```

**Comment by CrossEye** [comment]
LGTM

**Comment by davidchambers** [comment]
Let's consider the sequence of steps one must take in order to arrive at `R.thunk`:

1.  One wishes to apply `f` to zero or more arguments. Let's arbitrarily choose two as the number of arguments, to provide an example. Let's name those arguments `a` and `b`. So far so good. Applying functions to arguments is central to functional programming.

2.  There's a catch: `f` has side effects. We wish to avoid triggering the side effects right away, so we can't apply `f` immediately. This raises a red flag, but let's ignore it for now.

3.  There's a function for this: [`R.partial`][1]. `R.partial(f, a, b)` creates a function which, when invoked, will apply `f` to `a` and `b`. We're done!

4.  There's another catch: the function returned by `R.partial(f, a, b)` accepts any number of arguments to which `f` will *also* be applied. For some reason we're providing arguments to our "thunk", and we wish to ignore these arguments. Let's ignore *this* for now as well.

5.  The additional arguments to `f` would not be a problem if `f` were actually a binary function, but there's another catch: `f` is a variadic function!

To summarize, __when working with variadic functions with uncontrolled side effects, we run into problems if we provide more than *n* arguments to an *n*-ary function__. Rare circumstances indeed!

Not only are the circumstances rare, but they only arise if we do naughty things. First, functions should not have *uncontrolled* side effects. Perhaps `f` reads a file from disk, and we don't want to trigger this side effect until later. The correct solution is to use a Task (or similar type, but certainly *not* a Promise) to represent the delayed computation. Secondly, we should never apply an *n*-ary function to more than *n* arguments. I realize this is common practice in some circles, but it's mathematically nonsensical.

Let's not include a function which is only useful when one is naughty. Doing so would implicitly condone the behaviour. In this case, `R.partial(R.binary(f), a, b)` or `R.nAry(0, R.partial(f, a, b))` would be fine. They both look weird, which is a *good* thing: the weirdness may be a :warning: for readers.

:thumbsdown:


[1]: http://ramdajs.com/0.17/docs/#partial

**Comment by scott-christopher** [comment]
This doesn't necessarily need side effects to be useful. Consider a lazy right fold where you don't wish to evaluate the rest of the accumulating value if it is not needed.
```js
function foldRThunk(fn, acc, list) {
  return R.isEmpty(list) ? acc : fn(list[0], thunk(foldRThunk, fn, acc, R.tail(list)));
}
exists = (p, list) => foldRThunk((a, b) => p(a) || b(), false, list);
```
Having said that, I wasn't aware that `partial` already accepts the entire argument list to return a nullary function. I feel that renders the need for a specific `thunk` function a bit moot.

**Comment by davidchambers** [comment]
Thanks for mentioning laziness, @scott-christopher. I had not considered that use case. I agree, though, that `partial` and `() =>` are acceptable solutions.

**Comment by CrossEye** [comment]
It's not just a matter of the sort of examples described earlier, or of lazy data structures either.  There's simply the notion of computationally expensive functions many of which might be prepped ahead of time, but only one of which will end up running.  That, I suppose is another form of laziness, but is somewhat different from the lazy fold example.

This, of course, can be handled by fat arrows or by `partial`, but here is where I think I come down in favor of syntactic sugar when @davidchambers might not:  Even if a thunk can be created by partial application, it's the `thunk` abstraction I'm thinking of, not the `partial application` one.  A `thunk` is a delayed computation.  A `partial application` is the early application of a subset of the parameters of a function.  The fact that one can create a thunk by making that subset the full  complement of parameters does not make a thunk a partial application; it simply means that there is a way to implement it as such.  While it may sound like splitting hairs to some, it's a very real distinction to me.  It's the same reason that I think `head` should not be deprecated in favor of `nth(0)`.  They are separate abstractions which merely happen to coincide.

**Comment by scott-christopher** [comment]
My point was very much inline with using thunks for lazy evaluation (the lazy fold was just an example that I had within arms reach). I agree that `thunk` provides a clearer intent to the reader than `partial`, but I'm personally happy with using fat arrow syntax so I'm impartial as to whether we offer this or not.

**Comment by davidchambers** [comment]
While my preference is not to include this function, I don't mind if we do so. I leave it to you, @CrossEye.

**Comment by CrossEye** [comment]
@buzzdecafe?  @paldepind?  @TheLudd?  Any preferences?

I also don't have particularly strong feelings.

**Comment by TheLudd** [comment]
> I also don't have particularly strong feelings.

Me neither

**Comment by buzzdecafe** [comment]
abstain

**Comment by paldepind** [comment]
> I agree, though, that partial and () => are acceptable solutions.

I agree. `partial` and `() =>` are common and well understood.

**Comment by raine** [comment]
`R.thunk` doesn't seem like a compelling enough addition to Ramda so closing this.

**Comment by CrossEye** [comment]
**ker-Thunk**

**Pull request closed without merging**